### PR TITLE
Make operator TreeEnsemble 5x faster for batches of size 100.000

### PR DESCRIPTION
**Pull request by xadupre**
**Description**:

The fix only improves the operator. It parallelizes by trees instead of observations.

**1 target or 1 class**

BEFORE

image

AFTER

image

**multiregression or multiclass**

BEFORE

image

AFTER

image

**Motivation and Context**
This change makes onnxruntime as fast as scikit-learn on big batches.

### CODE REVIEW

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h [hunk 1]
```diff
@@ -266,11 +269,11 @@ void TreeEnsembleCommon<ITYPE, OTYPE>::ComputeAgg(concurrency::ThreadPool* ttp,
   if (n_targets_or_classes_ == 1) {
     if (N == 1) {
       ScoreValue<OTYPE> score = {0, 0};
-      if (n_trees_ <= parallel_tree_) {
+      if (n_trees_ <= parallel_tree_) { /* section A */
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h:272]: > section A [](start = 42, length = 10)

Comments that define the meaning of sections are helpful for reviews and readability

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h [hunk 2]
```diff
@@ -7,6 +7,9 @@
 #include "core/platform/ort_mutex.h"
 #include "core/platform/threadpool.h"
 
+#define TREEENSEMBLE_BATCHSIZE 256
+#define TREEENSEMBLE_MAXSIZE 8589934592
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h:11]: constexpr?

I am wondering why we are having so much code in the header? Are we sharing it with other .cc files?

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h [hunk 3]
```diff
@@ -284,46 +287,86 @@ void TreeEnsembleCommon<ITYPE, OTYPE>::ComputeAgg(concurrency::ThreadPool* ttp,
...
-            },
-            0);
+        agg.FinalizeScores1(z_data + i, score,
+                            label_data == nullptr ? nullptr : (label_data + i));
       }
+    } else { /* section D */
+      /*
+      // Parallelization by trees.
+      // This could use an array N * nth where nth is the number of threads.
+      // It would requires function omp_get_thread_num and omp_get_max_threads.
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h:308]: >requires [](start = 18, length = 8)

nit: require

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h [hunk 4]
```diff
@@ -284,46 +287,86 @@ void TreeEnsembleCommon<ITYPE, OTYPE>::ComputeAgg(concurrency::ThreadPool* ttp,
...
+      std::vector<ScoreValue<OTYPE>> scores_t(n_trees_ * N, {0, 0});
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h:309]: >scores_t [](start = 37, length = 8)

_t usually designates a type. score would be just fine

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h [hunk 5]
```diff
@@ -284,46 +287,86 @@ void TreeEnsembleCommon<ITYPE, OTYPE>::ComputeAgg(concurrency::ThreadPool* ttp,
...
+          [this, &scores_t, &agg, z_data, label_data, N](ptrdiff_t i) {
+            for (int64_t j = 1; j < this->n_trees_; ++j) {
+              agg.MergePrediction1(scores_t[i], scores_t[j * N + i]);
+            }
+            agg.FinalizeScores1(z_data + i, scores_t[i], label_data == nullptr ? nullptr : (label_data + i));
+          },
+          0);
+      */
+      auto num_threads = std::min<int32_t>(concurrency::ThreadPool::DegreeOfParallelism(ttp), SafeInt<int32_t>(n_trees_));
+      std::vector<ScoreValue<OTYPE>> scores(num_threads * N);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h:332]: > N); [](start = 57, length = 4)

Would you like to provide a default value as above?

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h [hunk 6]
```diff
@@ -284,46 +287,86 @@ void TreeEnsembleCommon<ITYPE, OTYPE>::ComputeAgg(concurrency::ThreadPool* ttp,
...
+          });
+
+      concurrency::ThreadPool::TrySimpleParallelFor(
+          ttp,
+          num_threads,
+          [&agg, &scores, num_threads, label_data, z_data, N](ptrdiff_t batch_num) {
+            auto work = concurrency::ThreadPool::PartitionWork(batch_num, num_threads, N);
+            for (auto i = work.start; i < work.end; ++i) {
+              for (int64_t j = 1; j < num_threads; ++j) {
+                agg.MergePrediction1(scores[i], scores[j * N + i]);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h:355]: > agg.MergePrediction1(scores[i], scores[j * N + i]); [](start = 15, length = 52)

Do we need a mutex here?

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h [hunk 7]
```diff
@@ -284,46 +287,86 @@ void TreeEnsembleCommon<ITYPE, OTYPE>::ComputeAgg(concurrency::ThreadPool* ttp,
...
 
-              agg.FinalizeScores1(z_data + i * n_targets_or_classes_, score,
-                                  label_data == nullptr ? nullptr : (label_data + i));
-            },
-            0);
+        agg.FinalizeScores1(z_data + i, score,
+                            label_data == nullptr ? nullptr : (label_data + i));
       }
+    } else { /* section D */
+      /*
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h:305]: >    /* [](start = 2, length = 6)

This code is commented out. Would you like to remove it?

**Comment by yuslepukhin** [comment]
>               std::lock_guard<OrtMutex> lock(merge_mutex);

Do we need a mutex here since each thread has its own block?

---
Refers to: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h:384 in ad600af. [](commit_id = ad600af6f2a6b45cf0735b86a2cfcb95352c88f2, deletion_comment = False)

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h (hunk 2 above)
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h:11]: constexpr?

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h (hunk 7 above)
Comment (xadupre) [reply onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h]: done

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h (hunk 2 above)
Comment (xadupre) [reply onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h:11]: not used any more, sorry for leaving them

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h (hunk 1 above)
Comment (xadupre) [reply onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h]: Done, I also used this section in two unit tests to check every case was tested.

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h (hunk 3 above)
Comment (xadupre) [reply onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h]: comment updated

File: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h (hunk 4 above)
Comment (xadupre) [reply onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h]: done

**Comment by xadupre** [comment]
> > ```
> >           std::lock_guard<OrtMutex> lock(merge_mutex);
> > ```
> 
> Do we need a mutex here since each thread has its own block? Can we we make code take advantage of it and rid of the mutex?
> 
> Refers to: onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h:384 in [ad600af](https://github.com/microsoft/onnxruntime/commit/ad600af6f2a6b45cf0735b86a2cfcb95352c88f2). [](commit_id = [ad600af](https://github.com/microsoft/onnxruntime/commit/ad600af6f2a6b45cf0735b86a2cfcb95352c88f2), deletion_comment = False)

Removed.

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Pull request merged**